File: R/algorithmic_toolbox/20_organizing_a_lottery.R
#!/usr/bin/env Rscript
ReadInputData <- function(conn) {
    nrSegments <- as.integer(unlist(strsplit(readLines(conn, n = 1), split = " ")))[1]
    segments <- matrix(0, nrow = nrSegments, ncol = 2)
    colnames(segments) <- c("left", "right")
    for (ix in 1:nrSegments) {
        segments[ix, ] <- as.integer(unlist(strsplit(readLines(conn, n = 1), split = " ")))
    }
    points <- as.integer(unlist(strsplit(readLines(conn, n = 1), split = " ")))
    list(nrSegments=nrSegments, segments=segments, points=points)
}

getNumberOfSuitableEndpoints <- function(endPoints, size, point) {
    BinarySearch <- function(a, b) {
        if (a == b) return(a)
        mid <- (a + b) %/% 2
        if (endPoints[mid] <= point) BinarySearch(mid + 1, b) else BinarySearch(a, mid)
    }
    if (endPoints[size] <= point) size else BinarySearch(1, size) - 1
}

CalcIntersectionSize <- function(point, sortedLeft, sortedNegatedRight, nrSegments) {
    intervalsFromLeft <- getNumberOfSuitableEndpoints(sortedLeft, nrSegments, point)
    intervalsFromRight <- getNumberOfSuitableEndpoints(sortedNegatedRight, nrSegments, -point)
    intervalsFromLeft + intervalsFromRight - nrSegments
}

NumberOfSegmentsContainingPoints <- function(endPoints, nrSegments, points) {
    sortedLeft <- sort(endPoints[,"left"])
    sortedNegatedRight <- sort(-endPoints[,"right"])
    sapply(points, CalcIntersectionSize, sortedLeft, sortedNegatedRight, nrSegments)
}

Main <- function() {
    conn <- file("stdin", "r")
    inputData <- ReadInputData(conn)
    close(conn)
    result <- NumberOfSegmentsContainingPoints(inputData$segments, inputData$nrSegments, inputData$points)
    cat(result, fill = TRUE)
}

if (!interactive()) {
    Main()
}

File: Julia/algorithmic_toolbox/20-organizing-a-lottery.jl
converttointlist(line) = map(x -> parse(Int, x), split(line))


function readintervals(nrsegments)
    segments = Array{Int}(undef, (nrsegments, 2))
    for ix in 1:nrsegments
        segments[ix, :] = readline() |> converttointlist
    end
    segments
end


function nrsuitableendpoints(endpoints, size, point)
    function search(a, b)
        if a == b
            return a
        end
        mid = (a + b) ÷ 2
        endpoints[mid] <= point ? search(mid + 1, b) : search(a, mid)
    end

    endpoints[end] <= point ? size : search(1, size) - 1
end


function intersectionsize(sortedleft, sortednegatedright, nrsegments, point)
    goodleftends = nrsuitableendpoints(sortedleft, nrsegments, point)
    goodrightends = nrsuitableendpoints(sortednegatedright, nrsegments, -point)
    goodleftends + goodrightends - nrsegments
end


function nrsegmentscontainingpoints(segments, nrsegments, points)
    sortedleft = sort(segments[:, 1])
    sortednegatedright = sort(-segments[:, 2])
    map(p -> intersectionsize(sortedleft, sortednegatedright, nrsegments, p), points)
end


function main()
    nrsegments, _ = readline() |> converttointlist
    segments = readintervals(nrsegments)
    points = readline() |> converttointlist
    result = nrsegmentscontainingpoints(segments, nrsegments, points)
    println(join(result, ' '))
end


main()
